### ISSUE TITLE
make check fails on test_vectors test with openssl 3.0

### ISSUE THREAD

**Issue by paride** [strongswan/strongswan#759]
**System (please complete the following information):**
 - OS: Ubuntu 22.04 (devel)
 - Kernel version (if applicable): 5.13.0-19-generic
 - strongSwan version(s): 5.9.4
 - Tested/confirmed with the latest version: yes

**Describe the bug**
The test suite (`make check`) fails to run `test_vectors` when strongswan is compiled against OpenSSL 3.0 (patched so that https://github.com/strongswan/strongswan/issues/753 is fixed). Log excerpt:
```
  Running suite 'vectors':
    Running case 'transforms': -++-+++++
      Failure in 'test_vectors': success: test vector for CAST_CBC from 'openssl' plugin failed (suites/test_vectors.c:48, i = 0)
      Failure in 'test_vectors': success: test vector for HASH_MD4 from 'openssl' plugin failed (suites/test_vectors.c:48, i = 3)
  Passed 0/1 'vectors' test cases
```

**To Reproduce**
Steps to reproduce the behavior. This is not exactly how the Ubuntu package is built, I tried to setup a reproducer that is easy to follow even without specific knowledge about Debian/Ubuntu packaging:

1. `lxc launch ubuntu-daily:jammy paride-j`
2. `lxc exec paride-j bash` (or start a Jammy system in your preferred way)
3. `echo 'deb-src http://archive.ubuntu.com/ubuntu jammy main' >> /etc/apt/sources.list`
4. `add-apt-repository ppa:schopin/openssl-3.0.0`
5. `apt update`
6. `apt full-upgrade`
7. `apt build-dep strongswan`
8. `git clone --depth 1 https://github.com/strongswan/strongswan`
9. `cd strongswan && ./autogen.sh`
10. `./configure --enable-addrblock --enable-agent --enable-bypass-lan --enable-ccm --enable-certexpire --enable-chapoly --enable-cmd --enable-ctr --enable-curl --enable-eap-aka --enable-eap-gtc --enable-eap-dynamic --enable-eap-identity --enable-eap-md5 --enable-eap-mschapv2 --enable-eap-peap --enable-eap-radius --enable-eap-tls --enable-eap-tnc --enable-eap-ttls --enable-error-notify --enable-forecast --enable-gcm --enable-gcrypt --enable-ha --enable-kernel-libipsec --enable-ldap --enable-led --enable-lookip --enable-mediation --enable-ntru --enable-openssl --enable-pkcs11 --enable-test-vectors --enable-tpm --enable-tss-tss2 --enable-unity --enable-xauth-eap --enable-xauth-pam --disable-blowfish --disable-fast --disable-des` (not all needed I'm sure)
10. `make`
11. `make check`

```
  Running suite 'vectors':
    Running case 'transforms': -++-+++++
      Failure in 'test_vectors': success: test vector for CAST_CBC from 'openssl' plugin failed (suites/test_vectors.c:48, i = 0)
      Failure in 'test_vectors': success: test vector for HASH_MD4 from 'openssl' plugin failed (suites/test_vectors.c:48, i = 3)
  Passed 0/1 'vectors' test cases
```

Cc: @pemensik

**Comment by tobiasbrunner** [comment]
See https://github.com/strongswan/strongswan/issues/753#issuecomment-966483799, but since the title of the other issue didn't mention these failure we can keep this as separate issue.

**Comment by paride** [comment]
Thanks for looking at this @tobiasbrunner, let me know if you need further feedback. FYI we're tracking the Ubuntu strongswan/openssl3 build issue here:

https://bugs.launchpad.net/strongswan/+bug/1946213

**Comment by tobiasbrunner** [comment]
I've pushed a commit to the _759-openssl-legacy_ branch that enables the "legacy" provider (optionally, it can be disabled to load only the default provider, or those configured via openssl.cnf).

**Comment by paride** [comment]
I can confirm that with a373b6aee36c01b7f6c5ca052c932595bc93d471 applied the test suite passes.

**Comment by pemensik** [comment]
Would it make sense to mark some ciphers required and some optional? Could it perhaps just warn() on MD{2,4,5} hashes, because their missing support is not a real issue I think? Is there way to extract more than number of algorithm?

I do not like legacy provider loaded by default. It contains crypto algorithms considered weak, I think they should stay unavailable unless administrator needs them explicitly. I would prefer disabling those algorithms instead of legacy provider enabled.

**Comment by tobiasbrunner** [comment]
> Could it perhaps just warn() on MD{2,4,5} hashes, because their missing support is not a real issue I think?

Might be possible, but there is currently no concept of algorithm strength beyond the filtering and ordering in the default IKE/ESP proposals (also see below).  I guess we could add something like that directly in the test suite if really necessary.

> I do not like legacy provider loaded by default. It contains crypto algorithms considered weak, I think they should stay unavailable unless administrator needs them explicitly. 

While we could disable the _load_legacy_ option, by default, it would break many user's existing remote access VPN because EAP-MSCHAPv2 is required, which depends on MD4 and DES-ECB. Not sure if Windows servers support EAP-MD5 (by default or at all). And server admins might have EAP-MSCHAPv2 configured for client authentication because most client implementations e.g. those by Apple or Microsoft support it out of the box (not sure if the latter support anything else if one doesn't want to use certificates for client authentication).

Just loading the provider shouldn't really be a problem because weak algorithms either can't be used for IKE/ESP anyway (e.g. MD2 or MD4) or they are not included in the default proposals (e.g. HMAC-MD5 or the uncommon ciphers like CAST or Blowfish), that is, users/admins have to explicitly configure proposals that list such algorithms to use them.

**Comment by paride** [comment]
@tobiasbrunner I am not fully familiar with the strongswan devel process / branching model. Do you plan to merge https://github.com/strongswan/strongswan/tree/759-openssl-legacy into master at some point, will you open a PR out of it, or something else entirely? Does the feedback label mean that you need further feedback before merging? In this case I'll be happy to help.

**Comment by tobiasbrunner** [comment]
> Do you plan to merge https://github.com/strongswan/strongswan/tree/759-openssl-legacy into master at some point,

Yes, absolutely, as something like it will be necessary for compatibility with OpenSSL 3 anyway. The next release is not planned for two months, though (see [Milestones](https://github.com/strongswan/strongswan/milestones)).

> will you open a PR out of it, or something else entirely?

I usually don't for my own branches (also, I generally rebase - even if I merge multiple commits, I rebase first). Once I push it to master I'll mark it with a "Closes" tag so the commit will close this issue and be associated with it.

> Does the feedback label mean that you need further feedback before merging?

I usually added it after responding to an issue (we had a "Feedback" state in the old issue tracker, to differentiate from new tickets), but since we now also have the "needs triage" label, that might actually be redundant. I guess your interpretation would make more sense, but for that a different name would be more helpful (like "awaiting feedback" or "feedback needed"). For now I'm going to remove the label. To be honest I'm still figuring this out myself :)

If you have any feedback, though, regarding the patch or the discussion above, it'd be appreciated.

**Comment by iamsais** [comment]
Could not load legacy MD4 even after updating from 5.9.2 to 5.9.9, there is no change in the error,

"_**EAP-MS-CHAPv2 failed, no MD4 hasher available**_"

Enabled by default: charon.plugins.openssl.load_legacy = yes

Should i configure anything other than this? Please help

**Comment by tobiasbrunner** [comment]
> With Openssl 3.0.5

On what platform/distribution? Did you build OpenSSL or strongSwan yourself? What does the log say during initialization (should be something like `"providers loaded by OpenSSL: ..."`)?

**Comment by iamsais** [comment]
Platform is Android (x86_64)

I have built both Openssl 3.0.5 (.so) separately and Android project builds the Strongswan while building the gradle project.

Log:
```
2023-02-11 21:25:51.732 8494-8609/com.versa.sase I/charon: 00[LIB] providers loaded by OpenSSL: default
2023-02-11 21:25:51.735 8494-8609/com.versa.sase I/charon: 00[LIB] loaded plugins: androidbridge charon android-log openssl fips-prf random nonce pubkey chapoly curve25519 pkcs1 pkcs8 pem xcbc hmac kdf socket-default revocation eap-identity eap-mschapv2 eap-md5 eap-gtc eap-tls x509
```

**Comment by tobiasbrunner** [comment]
Building OpenSSL 3's `libcrypto` as a shared library won't work as the legacy provider will be a separate shared module. Build it as a static library that gets linked into `libstrongswan`. I've recently pushed some changes to the _android-updates_ branch that include a build script for OpenSSL (I'm currently updating the documentation).
